[state_dict][9/N] Add get and set APIs for model and optimizer state_dict

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112414
* __->__ #112203
* #112167

The original get_state_dict and set_state_dict pair is too complicated because of the possible combinations of usages. This PR adds the APIs to get/set model_state_dict and optimizer_state_dict seperately.

Differential Revision: [D50713584](https://our.internmc.facebook.com/intern/diff/D50713584/)

fegin committed:
[state_dict][9/N] Add get_model_state_dict/get_optimizer_state_dict and set_model_state_dict/set_optimizer_state_dict

The original get_state_dict and set_state_dict pair is too complicated because of the possible combinations of usages. This PR adds the APIs to get/set model_state_dict and optimizer_state_dict seperately.

Differential Revision: [D50713584](https://our.internmc.facebook.com/intern/diff/D50713584/)

[ghstack-poisoned]

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+    options: Optional[StateDictOptions] = None,
+) -> Dict[str, ValueType]:
+    """
+    Return the model state_dict and optimizers state_dict.
```
nit: Return the model state_dict.

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+    """
+    Return the model state_dict and optimizers state_dict.
+
+    ``state_dict`` is a function that can process any module
```
nit: ``get_model_state_dict``

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+    that is parallelized by FSDP/fully_shard, DDP/replicate,
+    tensor_parallel/parallelize_module,
+    and any combination of these parallelisms. The main functions of
+    ``state_dict`` are: 1.) creating a model and optimizer state_dict that can
```
ditto.

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+    A canonical FQN refers to the FQN based on a parameter's position in an
+    nn.Module hierarchy. More specifically, a canonical FQN to a parameter is the
+    FQN returned by ``module.named_parameters()`` or ``module.named_buffers()``
+    when module is not distributed by any parallelisms. Since the optimizer
```
Nit: Remove the optimizer description.

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+        ddp_optim = torch.optim.Adam(model.parameters(), lr=1e-3)
+
+
+        ddp_state_dict, ddp_optim_state_dict = state_dict(ddp_model, ddp_optim)
```
nit: get_state_dict()

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+
+    Args:
+        model (nn.Module): the nn.Module to the model.
+        optimizers (Union[None, Optimizer, Iterable[Optimizer]]):
+            The optimizers that are used to optimize ``model``.
+        optim_only (bool): if optim_only is True, the returned model state_dict
+            will be empty (default: False)
```
Nit: remove.

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+            `StateDictOptions` for the details.
+
+    Returns:
+        ``Tuple`` that contain model state_dict and optimizer state_dict:
```
Nit: update return.

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -526,13 +526,188 @@ def _load_optim_state_dict(
+    options: Optional[StateDictOptions] = None,
+) -> Dict[str, ValueType]:
+    """
+    Return the model state_dict and optimizers state_dict.
```
ditto

fegin committed:
Update on "[state_dict][9/N] Add get and set APIs for model and optimizer state_dict"


The original get_state_dict and set_state_dict pair is too complicated because of the possible combinations of usages. This PR adds the APIs to get/set model_state_dict and optimizer_state_dict seperately.

Differential Revision: [D50713584](https://our.internmc.facebook.com/intern/diff/D50713584/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][9/N] Add get and set APIs for model and optimizer state_dict"


The original get_state_dict and set_state_dict pair is too complicated because of the possible combinations of usages. This PR adds the APIs to get/set model_state_dict and optimizer_state_dict seperately.

Differential Revision: [D50713584](https://our.internmc.facebook.com/intern/diff/D50713584/)

[ghstack-poisoned]

wz337 commented on torch/distributed/checkpoint/state_dict.py:
```diff
@@ -74,32 +74,34 @@ def gc_context():
+    - ``keep_submodule_prefixes``: when ``submodules`` is not None, this option
+      indicates whether to keep the submodule prefixes from the state_dict keys.
+      or example, if the submodule is ``module.pretrain`` and the full FQN of
+      the parameter is ``pretrain.layer1.weight`` of the param. When this option
+      is True, the parameter's key in the returned state_dict will be
+      ``layer1.weight``. If the options is False, the key will be
+      ``pretrain.layer1.weight``.
```
Just for my understanding, keep_submodule_prefixes=True will return ``layer1.weight`` and keep_submodule_prefixes=False will return ``pretrain.layer1.weight``. I thought it was the other way around.

wz337 reviewed (approved):
LGTM!

fegin committed:
Update on "[state_dict][9/N] Add get and set APIs for model and optimizer state_dict"


The original get_state_dict and set_state_dict pair is too complicated because of the possible combinations of usages. This PR adds the APIs to get/set model_state_dict and optimizer_state_dict seperately.

Differential Revision: [D50713584](https://our.internmc.facebook.com/intern/diff/D50713584/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge
